precompiles/xcm/src/lib.rs

```diff
@@ -114,130 +87,94 @@ where
+        .into();
+        Self::assets_withdraw_v1_internal(
+            handle,
+            assets.into(),
+            amounts.into(),
+            beneficiary,
+            is_relay,
+            parachain_id,
+            fee_index,
+        )
     }
-}
 
-/// The supported beneficiary account types
-enum BeneficiaryType {
-    /// 256 bit (32 byte) public key
-    Account32,
-    /// 160 bit (20 byte) address is expected
-    Account20,
-}
+    #[precompile::public("assets_withdraw(address[],uint256[],address,bool,uint256,uint256)")]
+    fn assets_withdraw_evm_v1(
+        handle: &mut impl PrecompileHandle,
+        assets: BoundedVec<Address, GetMaxAssets<Runtime>>,
+        amounts: BoundedVec<U256, GetMaxAssets<Runtime>>,
+        recipient_account_id: Address,
+        is_relay: bool,
+        parachain_id: U256,
+        fee_index: U256,
+    ) -> EvmResult<bool> {
+        let beneficiary: Junction = Junction::AccountKey20 {
+            network: None,
+            key: recipient_account_id.0.to_fixed_bytes(),
+        }
+        .into();
+        Self::assets_withdraw_v1_internal(
+            handle,
+            assets.into(),
+            amounts.into(),
+            beneficiary,
+            is_relay,
+            parachain_id,
+            fee_index,
+        )
+    }
 
-impl<Runtime, C> XcmPrecompile<Runtime, C>
-where
-    Runtime: pallet_evm::Config
-        + pallet_xcm::Config
-        + orml_xtokens::Config
-        + pallet_assets::Config
-        + AddressToAssetId<<Runtime as pallet_assets::Config>::AssetId>,
-    <<Runtime as frame_system::Config>::RuntimeCall as Dispatchable>::RuntimeOrigin:
-        From<Option<Runtime::AccountId>>,
-    <Runtime as frame_system::Config>::AccountId: Into<[u8; 32]>,
-    <Runtime as frame_system::Config>::RuntimeCall: From<pallet_xcm::Call<Runtime>>
-        + From<orml_xtokens::Call<Runtime>>
-        + Dispatchable<PostInfo = PostDispatchInfo>
-        + GetDispatchInfo,
-    XBalanceOf<Runtime>: TryFrom<U256> + Into<U256> + From<u128>,
-    <Runtime as orml_xtokens::Config>::CurrencyId:
-        From<<Runtime as pallet_assets::Config>::AssetId>,
-    C: Convert<MultiLocation, <Runtime as pallet_assets::Config>::AssetId>,
-{
-    fn assets_withdraw_v1(
+    fn assets_withdraw_v1_internal(
         handle: &mut impl PrecompileHandle,
-        beneficiary_type: BeneficiaryType,
-    ) -> EvmResult<PrecompileOutput> {
-        let mut input = handle.read_input()?;
-        input.expect_arguments(6)?;
+        assets: Vec<Address>,
+        amounts: Vec<U256>,
+        beneficiary: Junction,
+        is_relay: bool,
+        parachain_id: U256,
+        fee_index: U256,
+    ) -> EvmResult<bool> {
+        // Check that assets list is valid:
+        // * all assets resolved to multi-location
+        // * all assets has corresponded amount
+        if assets.len() != amounts.len() || assets.is_empty() {
+            return Err(revert("Assets resolution failure."));
+        }
```

Comment:
Not sure if it's still correct to do this at the beginning - on line 159 you use `filter_map`, which might reduce the size of the `assets` vector. The `zip` used later might end up discarding some values, and mixing up the asset Ids with amounts.

Reply:
Good catch, I'll move it below